src/plugins/ArithmeticExceptions.cpp

```diff
@@ -0,0 +1,366 @@
+// ArithmeticExceptions.h (Oclgrind)
+// Copyright (c) 2016, Moritz Pflanzer,
+// Imperial College London. All rights reserved.
+//
+// This program is provided under a three-clause BSD license. For full
+// license terms please see the LICENSE file distributed with this
+// source code.
+
+// The recommendations how to detect undefined behaviour haven been adopted from:
+// https://www.securecoding.cert.org/confluence/display/c/INT32-C.+Ensure+that+operations+on+signed+integers+do+not+result+in+overflow
+// https://www.securecoding.cert.org/confluence/display/c/INT33-C.+Ensure+that+division+and+remainder+operations+do+not+result+in+divide-by-zero+errors
+// Overflow in unary minus operator cannot be checked in LLVM IR
+// Oversize shifts are well-defined in OpenCL
+// Arithmetic operations involving vector types do not seem to produce nsw or nuw flags. This means overflows cannot be detected.
+
+#include "core/common.h"
+#include "core/Context.h"
+#include "core/WorkItem.h"
+
+#include <sstream>
+#include <cfloat>
+
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Type.h"
+
+#include "ArithmeticExceptions.h"
+
+#include "core/Kernel.h"
+#include "core/KernelInvocation.h"
+
+using namespace oclgrind;
+using namespace std;
+
+#define HALF_MAX 65504
+
+static long getSignedMinValue(const unsigned int bits)
+{
+    return -(1L << (bits - 2)) - (1L << (bits - 2));
+}
+
+static long getSignedMaxValue(const unsigned int bits)
+{
+    return ((1L << (bits - 2)) - 1L) + (1L << (bits - 2));
+}
+
+static long getUnsignedMaxValue(const unsigned int bits)
+{
+    return ((1L << (bits - 1)) - 1L) + (1L << (bits - 1));
+}
+
+void ArithmeticExceptions::instructionExecuted(
+  const WorkItem *workItem, const llvm::Instruction *instruction,
+  const TypedValue& result)
+{
+    switch(instruction->getOpcode())
+    {
+        case llvm::Instruction::Add:
+        {
+            const auto&& BinOp = llvm::dyn_cast<llvm::BinaryOperator>(instruction);
+
+            // Check for signed overflow
+            // Report an exception iff LLVM would create a poisoned value on overflow
+            // Other than that it is not possible to differentiate between signed/unsigned values
+            if(BinOp->hasNoSignedWrap())
```

Comment:
No, the plugin will only warn about signed wrap. Unsigned wrap is well defined in OpenCL (and C99).